### PR TITLE
Early implementation of "implementation-only imports"

### PR DESCRIPTION
**Pull request by jrose-apple**
I'm going to implement them with a terrible underscored name and no correctness checking for now.

rdar://problem/48985979

### CODE REVIEW

**Comment by jrose-apple** [comment]
(at this point I don't even have the implementation, just a refactoring I want to test against LLDB)

**Comment by jrose-apple** [comment]
apple/swift-lldb#1139

@swift-ci Please test

**Comment by jrose-apple** [comment]
Timeout in an LLDB test -> probably not related. Clear to move forward!

**Comment by jrose-apple** [comment]
apple/swift-lldb#1139

@swift-ci Please test

**Comment by jrose-apple** [comment]
TODO list of things I can think of off the top of my head:

- LLDB support (debugging FooKit vs. debugging a client of FooKit)
- Restrictions on use:
  - inlinable code, including default arguments
  - types in public and `@usableFromInline` entities
  - stored properties in non-resilient types (including classes) – we probably need a better story for this
- Conformances need to be allowed but not cause problems when we try to deserialize them. (That's currently not the case because we assume conformances might be needed to form the types of stored properties.)
  - This one gets easier if we limit implementation-only imports to only support textual interfaces (because we could just not print them), but textual interfaces require resilience, and that means they can't be used elsewhere, which would be unfortunate.
- How does this interact with `@testable import`?
- Needs more tests.
  - Needs a test for when a module uses `@_uncheckedImplementationOnly` in one file and nothing or `@_exported` in another.

I'm planning to write up a pseudo-proposal soon, in preparation for possibly handing off this feature to someone else.

**Comment by jrose-apple** [comment]
Just rebasing, still need to write tests. John's owning this now but I can help out by making sure this is in shape.

@swift-ci Please test

**Comment by jrose-apple** [comment]
Ugh. I thought I copied the LLDB PR too.

apple/swift-lldb#1139
@swift-ci Please test

**Comment by jrose-apple** [comment]
Okay, I think this is in a reviewable state now. Most of the TODOs still apply, but it has tests, at least!

apple/swift-lldb#1139
@swift-ci Please test

**Comment by jrose-apple** [comment]
apple/swift-lldb#1139
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
apple/swift-lldb#1139
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@brentdax, would you remind reviewing this too? Both so it can get merged sooner rather than later, and to increase the number of people who understand this feature. (That means ask questions if something doesn't make sense.)

I do recommend doing the two commits separately; the first is a lot of updating call sites.

**Comment by jrose-apple** [comment]
apple/swift-lldb#1139
@swift-ci Please test

**Review by beccadax — APPROVED** [review]
I have a few suggestions for style and clarity, but they're all suggestions—nothing mandatory.

File: test/NameBinding/import-implementation-only.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend -emit-module -o %t/abcde.swiftmodule %S/Inputs/abcde.swift
+// RUN: %target-swift-frontend -emit-module -o %t/Library.swiftmodule %s -I %t
+
+// RUN: echo 'import Library; foo()' > %t/main.swift
```
[the remark this answers is not in the document]
Comment (jrose-apple) [reply test/NameBinding/import-implementation-only.swift]: Gotta save inodes!

More seriously, I don't like having to jump around between files to find out what a test is doing, so if something really is this small I'd rather have it inline.

File: include/swift/Basic/OptionSet.h [hunk 2]
```diff
@@ -87,6 +87,16 @@ class OptionSet {
     return !static_cast<bool>(set - *this);
   }
 
+  /// Check if two option sets contain the exact same options.
+  friend bool operator==(OptionSet lhs, OptionSet rhs) {
```
[the remark this answers is not in the document]
Comment (jrose-apple) [reply include/swift/Basic/OptionSet.h]: :-/ Yeah, I guess so.

**Comment by jrose-apple** [comment]
[Full macOS build here](https://ci.swift.org/job/swift-PR-osx/12103/), [full Linux tests passed](https://ci.swift.org/job/swift-PR-Linux/12104/)

Updated for Brent's feedback, and also a tiny bit of debugging support (enough for the test cases in apple/swift-lldb#1414).

**Comment by jrose-apple** [comment]
apple/swift-lldb#1139
@swift-ci Please smoke test

**Review by beccadax — APPROVED** [review]
Looks great now!

**Comment by jrose-apple** [comment]
Hm, I don't *think* those LLDB failures would come from this PR, but maybe?

apple/swift-lldb#1139
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
Why am I not seeing these LLDB failures locally?

**Comment by jrose-apple** [comment]
Maybe…

apple/swift-lldb#1139
@swift-ci Please clean test macOS

**Comment by jrose-apple** [comment]
\*sigh\* Well, I'm not sure whether the clean test did it, or whether something just got fixed (or disabled) on master, but it looks like we're clear to move forward. Merging.

**Pull request merged**